sky/server/server.py

```diff
@@ -287,23 +287,25 @@ async def validate(validate_body: payloads.ValidateBody) -> None:
     # these into a single call or have a TTL cache for (task, admin_policy)
     # pairs.
     logger.debug(f'Validating tasks: {validate_body.dag}')
-    try:
-        dag = dag_utils.load_chain_dag_from_yaml_str(validate_body.dag)
+
+    def validate_dag(dag: dag_utils.dag_lib.Dag):
         # TODO: Admin policy may contain arbitrary code, which may be expensive
         # to run and may block the server thread. However, moving it into the
         # executor adds a ~150ms penalty on the local API server because of
         # added RTTs. For now, we stick to doing the validation inline in the
         # server thread.
         dag, _ = admin_policy_utils.apply(
             dag, request_options=validate_body.request_options)
```

Comment:
Maybe in the future: Wondering if we could make our admin policy API to be async, so as to make it more efficient. Currently, this validate_dag function is blocking without cooporative, and putting it in running loop won't help that much? Did this resolve the issue for validate blocking the uvicorn?

Reply:
Yes, it runs in a separate thread executor and the thread executor is awaited in the uvicorn event loop

Reply:
> Wondering if we could make our admin policy API to be async

Yes, this can be incrementally. The general idea is we await async function in handler and convert sync function to async by wrapping it with a thread executor.

To this specific case of admin policy, user facing changes will also be involved, a new async interface should be added in addition to the synced `AdminPolicy.validate_and_mutate()`, and user should implement the async one if they want to apply admin policy more efficient.